### DISCUSSION TITLE
assigning properties to each elements (for drainage/Porousflow)

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#18542] [Q&A Modules: General] [answered]
Hi all,

I'd like to assign the particular value of drainage to each element instead of each block/surface.

by the value of drainage, I mean C in `PorousFlowPiecewiseLinearSink`
https://mooseframework.inl.gov/modules/porous_flow/boundaries.html

I checked https://mooseframework.inl.gov/modules/porous_flow/heterogeneous_models.html
but it assigns permeability base on the location (x-y), I want it to be based on the elementID.

It would be ideal if I can import value from like csv file into moose like (elementID, property value)
Not sure if this's possible?

Thank you guys.

Kind regards,
Traiwit

**Comment by WilkAndy** [comment]
Would DiracKernels help?

**Reply by Traiwit** [reply]
I tried to look into this, it doesn't look like what I want. I checked, most of DiracKernels specify at the location (x,y,z).

Let's step back to a more generic question, is it possible for MOOSE to assign material properties base on elementID instead of blockID? 

Kind regards,
Traiwit

**Reply by cpgr** [reply]
I don't think it is. Can you explain more why you can't use (x,y,z) coordinates for this?

**Reply by Traiwit** [reply]
It's just easier for us to integrate MOOSE to our framework, since we normally assigning properties base on elementID via Abaqus.

Question: if we assign properties base on location using data_file: `PiecewiseMultilinear` `PiecewiseMulticonstant`, where would the properties be assigned to? at the node of at the element?

Thanks you.

Kind regards,
Traiwit

**Reply by josebastiase** [reply]
Hi Traiwit,

Material properties are assigned to the elements and you can easily get the coordinates of the elementID when you generate the mesh or in paraview. 

You can assign spatial distributed properties with the workflow Andy and I made a while ago. I made a simple example for you. Seems to work fine.

https://drive.google.com/drive/folders/1zneEcLT-LpScxtw4HUdKI1_akEwoXauI?usp=sharing

Cheers,
Jose

**Reply by Traiwit** [reply]
by 'the coordinates of the elementID' you mean the centroid of the element right?

Thank you for the link, I will have a look tomorrow and get back to you.

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
Hi @josebastiase 

I checked the files, they are nice and highly likely to be applicable to my problem.

I have a quick question regarding the Python script, say if I have a CSV file with drainage value and its location like this >>> (x, y, z, drainage)

how should I reformat it so MOOSE can read it using the block below?

```
[C_fcn]
    type = PiecewiseMulticonstant
    direction = 'left right'
    data_file = C.data
  []
```

Thanks again.

Kind regards,
Traiwit

**Reply by cpgr** [reply]
Take a look at https://mooseframework.inl.gov/source/functions/PiecewiseMulticonstant.html

and some examples https://github.com/idaholab/moose/tree/next/test/tests/functions/piecewise_multilinear

**Reply by Traiwit** [reply]
@cpgr sorry for being stubborn, but how should I work around this if I have unstructured mesh (like the one below)? It seems like the examples only show cartesian data file

![image](https://user-images.githubusercontent.com/79562439/128975410-ae36bb31-899a-4d86-926d-5a1d57d84792.png)

![image](https://user-images.githubusercontent.com/79562439/128975217-6f400a6b-8195-4ab2-bd2c-b2fb99222374.png)

From this link: https://mooseframework.inl.gov/modules/porous_flow/heterogeneous_models.html

I guess I need to use `em2ex` (https://github.com/cpgr/em2ex) then, not sure what it is, but ill give it a go.

> This pre-processing must be performed using software that can take a geological model and convert it into a form that PorousFlow can read. One example of a pre-processing code that may be used is em2ex. This code converts a reservoir model in either Eclipse format or cell centered csv format to an Exodus mesh that can used in PorousFlow.

Please suggest if there is an alternative way to work with the problem.

Thank you.

Kind regards,
Traiwit

**Reply by cpgr** [reply]
I assume that each cell has a different property value, rather than each cell in the red domain having the same property?

In that case, I'm not sure how you can easily achieve this in MOOSE without making a custom AuxKernel or Material. And I don't think em2ex will help - it is made for going from a Petrel model to an exodus model. 

Outside writing a custom AuxKernel/Material,I have used [meshio](https://github.com/nschloe/meshio) before to quickly modify meshes before using in MOOSE. It reads Abaqus files and exports Exodus, so it might be possible to read the Abaqus mesh, assign the property values based on element id (if these aren't read directly), and export to exodus. Then you can read in those values as AuxVariables, and proceed as normal?

**Reply by cpgr** [reply]
Or maybe PiecewiseMultiConstant works with unstructured grids if you give it centroids? I don't know, but it might be worth trying first?

**Reply by Traiwit** [reply]
Sorry for making it confusing, that 2 colours are separated by a fault, which is a completely different problem, it should be like this instead

![image](https://user-images.githubusercontent.com/79562439/128980121-231ac9cb-331a-4f87-ac76-86518cf86df0.png)

Sure, I will have a look at both meshio and `PiecewiseMultiConstant`, I suspect `PiecewiseMultiConstant` accept same kind of data_file, but will look further 

Thank you @cpgr 

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
@WilkAndy I see this `VectorPostprocessorPointSource` under `DiracKernels `, seems like it might help


example from (test/tests/dirackernels/vectorPostprocessor_point_source/3d_vpp.i)

it read using `CSV reader`
![image](https://user-images.githubusercontent.com/79562439/129142905-6ea1a808-ec74-47ca-a847-0b621657bce8.png)
then `VectorPostprocessorPointSource`
![image](https://user-images.githubusercontent.com/79562439/129142882-7ef8ef34-e4cc-48ce-9d5c-7a5743b68937.png)
below is the example of the CSV input file
![image](https://user-images.githubusercontent.com/79562439/129142935-6c902c6d-6061-442a-98f6-3167186a97d1.png)

Not sure what does `DiracKernels ` do in general since it's not documented.

Kind regards,
Traiwit

**Reply by Traiwit** [reply]
Hi guys, just wonder what does `VectorPostprocessorPointSource` do to the variable? 

Can I instead use `VectorPostprocessorPointSource` to store what I ready from the CSV file to Auxvariable instead, so then I can assign the values to `material properties` OR `flux_function`

@WilkAndy do you have any suggestions on this?

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
Hi Traiwit,

Did you try it? It should be fairly easy to set up.

I think the cleaner way to do this would be with a SpatialUserObjectAux, however, there is currently no UserObject I can think of that has the right method for you. 

For now, my recommendation would be to use a FunctionAux with a PiecewiseMultiConstant function.
https://mooseframework.inl.gov/source/auxkernels/FunctionAux.html
https://mooseframework.inl.gov/source/functions/PiecewiseMulticonstant.html

Guillaume

**Reply by Traiwit** [reply]
Hi @GiudGiud, first I thought I'd be able to use `PiecewiseMultiConstant `, but when I checked the example, it seems like it only read the cartesian mesh value. Most of the data_file in the examples look like this (attached below)

![image](https://user-images.githubusercontent.com/79562439/129312776-f26c96eb-0d2b-49e2-9829-5c8c09fd9f77.png)

Just wondering, is it possible for `PiecewiseMulticonstant` to read unstructured grid data from CSV file? for example (x,y,z,value)

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
Ah good point!
For the small mesh you are using though you could get around that by expanding the grid, eg, having every single coordinate on every axis. 

I ll keep thinking about it but you may have identified a feature we are missing.

**Reply by Traiwit** [reply]
@GiudGiud if you look at my reply above, `VectorPostprocessorPointSource ` seem to have the most potential to read unstructured data, but seem like it stores the value into a variable instead of Auxvariable.

**Reply by GiudGiud** [reply]
Yes I just looked at the code and it's really not doing what you want. It's meant to create a source term in a PDE.
 
It could however very easily be turned into an AuxKernel. Let's see if others suggest better solutions, otherwise you, or one of us if you are not comfortable with doing this, could add that object to the framework.

**Reply by lynnmunday** [reply]
I just changed the ```VectorPostprocessorPointSource``` to ```ReporterPointSource```    So now that dirac kernel takes a reporter instead of a vpp and the syntax is slightly different.  I wanted to change it before people started using it.

**Comment by Traiwit** [comment]
@GiudGiud i agree, source term in a PDE isn't what I want, I want something that I can impose to material properties or BCs

**Reply by Traiwit** [reply]
@GiudGiud is there any update on this issue? 

A quick question, is there a read_file.C or (data_file.C) file or something that I can look into? I checked `PiecewiseMultiConstant.C` but couldn't find that data_file option in there.

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
No new object to do that afaik.
I m still convinced best way would be to add a spatialValue routine to the CSVReader then just use that in a SpatialUOAux

**Comment by ngrilli** [comment, marked as answer]
Dear @Traiwit 

Whenever I want to assign a material property based on element ID, 
what I do is mimicking the approach used for the crystal plasticity code.

You can use the UserObject:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/userobjects/ElementPropertyReadFile.C

to read a file of properties element by element.
If you have a structured mesh, the element ID will grow along the x axis, then y axis, then z axis.

Then you can use the user object into a material object, like in:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/crystal_plasticity/ComputeElasticityTensorCP.C

This is the way you can obtain your property based on element ID:
_read_prop_user_object->getData(_current_elem, 0);

An example input file is:
https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/monolithic_material_based/crysp_user_object.i

In the case of crystal plasticity, this method allows to assign Euler angles based on element ID,
but you can use the same for any other material property.

Trust this helps,
Nicolò

**Reply by Traiwit** [reply]
Thank you @ngrilli I believe this is very close to what I want.

I have some questions, maybe @GiudGiud can also help me out as well.

1. I check the examples I see that only crystal plasticity related files have `_read_prop_user_object->getData(_current_elem, 0);` in their .C file, does this mean I have to add this myself to other material properties .C files such as `PorousFlowPermeabilityConst` and `PorousFlowPorosityConst`

2. is it possible to convert what I read from `ElementPropertyReadFile` into function format? one of my goals is to impose the property's value to flux_function in `PorousFlowPiecewiseLinearSink`

```
          [./water_grad_right]
            type = PorousFlowPiecewiseLinearSink
          variable = porepressure
          boundary = 'right'
          pt_vals = '0 1e9'
          multipliers = '0 1e9' 
          flux_function = xxxx ******HERE*******
          fluid_phase = 0
          [../]
```

Thank you gents

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
1. Porous flow includes the tensor_mechanics module so you use ElementPropertyReadFile there.
It is indeed currently not used in porous_flow, but that does not mean you can't use it the same way it s used for crystal plasticity
2. It's possible if you code a new function. This new function will look at the position, get the element from the mesh, then call this _read_prop_user_object->getData(_current_elem, 0);

**Reply by Traiwit** [reply]
@GiudGiud thank you for your reply, i'm working on the first part, and I'm getting there :)

a quick question on the 2nd point, is there an example of creating 'a new function'?

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
I dont think there's an example. I would just copy-paste an existing one then modify it. For what you want you could start from SolutionFunction.C. There might be better starting points but this one seems close enough
EDIT: I had a look, the tutorial covers kernels, materials and auxvariables, but not functions. You can still have a look at tutorial 1 if you have not

**Reply by Traiwit** [reply]
@GiudGiud Thank you

with my limited knowledge of C++, I'm slowly working on it (still failing to compile)

I will come back as ask if I'm really stuck.

Kind regards,
Traiwit

**Comment by Traiwit** [comment]
Hi guys,

I'm having a problem converting the type of input parameter. I'm trying to follow what @ngrilli suggested https://github.com/idaholab/moose/discussions/18542#discussioncomment-1211668.

In the example, the code read `ElementPropertyReadFile` into `RealVectorValue`

but what I want now is to convert what I read in `ElementPropertyReadFile` to `RealTensorValue` (shown below)

It compiles fine, but when I run the input file I have this problem >

```
*** ERROR ***
The parameter "permeability" is being retrieved before being set.
```
I notice that what I read from .CSV isn't in tensor format, I've also tried to convert the csv cell into tensor (X 0 0 0 X 0 0 0 X) but then `ElementPropertyReadFile` won't read.

![image](https://user-images.githubusercontent.com/79562439/130902281-5f1747bf-8130-475b-8bd8-df8c8c165d88.png)

Please suggest if you know how to work with this issue, sorry for my limited knowledge of C++, hope I could explain it better.

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
what's in your input for this block?
permeability seems declared and retrieved properly here,

**Reply by Traiwit** [reply]
Hi @GiudGiud, that is when i didnt give permeability value into the block to it, 

![image](https://user-images.githubusercontent.com/79562439/130906147-0995698a-d1a9-41e4-a464-3bd182ea49bb.png)

but when I do, I have this error instead

```
  [./permeability_top]
    type = PorousFlowPermeabilityConstTC
    permeability = '1e-10 0 0   0 1e-10 0   0 0 1e-10'
    read_prop_user_object = prop_read
  [../]
```
![image](https://user-images.githubusercontent.com/79562439/130906207-f83d0fe4-61b2-4267-97c3-7cac3de5ce24.png)

I notice that permeability is <RealTensorValue> (x 0 0 0 x 0 0 0 x), but I only supply a single number into   `_permeability_qp[_qp]` 

Hence the error occurs, but I'm not sure how to fix this

**Reply by GiudGiud** [reply]
If you retrieve a parameter with getParam it needs to be set in the input file, or it needs a default in the validParams

Your error is a hypre error. Let's see if @fdkong has seen it before.

**Reply by Traiwit** [reply]
@GiudGiud noted, i fact i don't think i even need `_input_permeability` since I'm trying to obtain this value from `_read_prop_user_object` instead. So I just quoted it out, and it still complies.

Now the problem should be that RHS and LHS of this line are not match
![image](https://user-images.githubusercontent.com/79562439/130907412-fd55b2d6-18f0-4ba1-b17a-298034fce7f5.png)
since LHS should be tensor but the RHS is a single constant value (from CSV)

![image](https://user-images.githubusercontent.com/79562439/130907248-f0ed8fae-95d0-4c53-b597-09438f3b7677.png)

**Reply by GiudGiud** [reply]
Is there another routine than getData you could use for that object?
If not you can always fill the tensor component by component

**Reply by Traiwit** [reply]
Hi @GiudGiud,

I finally got it to works now, thank you for going through this with me

![image](https://user-images.githubusercontent.com/79562439/131072542-142d8042-6a69-4ff6-8cbe-faf04f234136.png)

Now I still need to import what I read from CSV via `ElementPropertyReadFile ` to Function format so then I can apply it to `PorousFlowPiecewiseLinearSink`

I tried `SolutionFunction.C` but it takes different type of userobject (SolutionUserObject rather than ElementPropertyReadFile)

Kind regards,
Traiwit
